### PR TITLE
Field ops features

### PR DESCRIPTION
**Pull request by gdeskos**
Added two features to field_ops
1) normalize --> Normalizes a field using its magnitude
2) global_max_magnitude --> Computes the global maximum of a field's magnitude

-- No unittests  --

### CODE REVIEW

File: amr-wind/core/field_ops.H [hunk 1]
```diff
@@ -293,36 +293,109 @@ inline void lincomb(
...
+
+/** Computes the global maximum of a field from all levels
+ * \ingroup field_ops
+ *
+ * \param [in] field we need to compute its global maximum magnitude
+ */
+template <typename FType>
+inline amrex::Real global_max_magnitude(FType& field)
+{
+    using namespace amrex;
```
Comment (sayerhs) [inline amr-wind/core/field_ops.H:336]:
```suggestion

```

File: amr-wind/core/field_ops.H [hunk 2]
```diff
@@ -293,36 +293,109 @@ inline void lincomb(
...
+    const auto& repo = field.repo();
+    const auto ncomp = field.num_comp();
 
+    amrex::Real maxglobal;
+    const int nlevels = repo.num_active_levels();
+    for (int lev = 0; lev < nlevels; ++lev) {
+        amrex::Real maxglobal = 0.0;
```
Comment (sayerhs) [inline amr-wind/core/field_ops.H:343]: This `maxglobal` is shadowing the outer variable. Has this code been tested?

File: amr-wind/core/field_ops.H [hunk 3]
```diff
@@ -293,36 +293,109 @@ inline void lincomb(
...
+        maxglobal = amrex::ReduceMax(
+            field(lev), 0.,
+            [=] AMREX_GPU_HOST_DEVICE(
+                Box const& b, Array4<Real const> const& field_arr) -> Real {
+                Real mx = 0.0;
```
Comment (sayerhs) [inline amr-wind/core/field_ops.H:348]:
```suggestion
            [=] AMREX_GPU_HOST_DEVICE(
                amrex::Box const& b, amrex::Array4<Real const> const& field_arr) -> amrex::Real {
                amrex::Real mx = 0.0;
```

File: amr-wind/core/field_ops.H [hunk 4]
```diff
@@ -293,36 +293,109 @@ inline void lincomb(
...
+                amrex::Loop(b, [=, &mx](int i, int j, int k) noexcept {
+                    Real Mag = 0.0;
+                    for (int icomp = 0; icomp < ncomp; ++icomp) {
+                        Mag = Mag + field_arr(i, j, k, icomp) *
+                                        field_arr(i, j, k, icomp);
+                    }
```
Comment (sayerhs) [inline amr-wind/core/field_ops.H:354]:
```suggestion
                    amrex::Real mag = 0.0;
                    for (int icomp = 0; icomp < ncomp; ++icomp) {
                        mag = mag + field_arr(i, j, k, icomp) *
                                        field_arr(i, j, k, icomp);
                    }
```

File: amr-wind/core/field_ops.H [hunk 5]
```diff
@@ -293,36 +293,109 @@ inline void lincomb(
...
+                    mx = amrex::max(std::sqrt(Mag), mx);
```
Comment (sayerhs) [inline amr-wind/core/field_ops.H:355]:
```suggestion
                    mx = amrex::max(std::sqrt(mag), mx);
```

File: amr-wind/core/field_ops.H [hunk 6]
```diff
@@ -293,36 +293,109 @@ inline void lincomb(
...
+                });
+                return mx;
+            });
+    }
+    ParallelAllReduce::Max<Real>(maxglobal, ParallelContext::CommunicatorSub());
```
Comment (sayerhs) [inline amr-wind/core/field_ops.H:360]:
```suggestion
    amrex::ParallelAllReduce::Max<amrex::Real>(maxglobal, amrex::ParallelContext::CommunicatorSub());
```

File: amr-wind/core/field_ops.H [hunk 7]
```diff
@@ -293,36 +293,109 @@ inline void lincomb(
...
+    return maxglobal;
+}
+
+/** Normalizes a field using its magnitude
+ * \ingroup field_ops
+ *
+ * \param [in] field that needs to be normalized
+ */
+template <typename FType>
+inline void normalize(FType& field)
```
Comment (sayerhs) [inline amr-wind/core/field_ops.H:370]:
```suggestion
inline void normalize(FType& field, const amrex::Real eps = 1.0e-12)
```

File: amr-wind/core/field_ops.H [hunk 8]
```diff
@@ -293,36 +293,109 @@ inline void lincomb(
...
+
+            amrex::ParallelFor(
+                bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                    // Compute magnitude
+                    amrex::Real mag = 0.;
+                    for (int icomp = 0; icomp < ncomp; ++icomp) {
+                        mag = mag + field_arr(i, j, k, icomp) *
+                                        field_arr(i, j, k, icomp);
+                    }
+                    if (mag > 1e-12) {
```
Comment (sayerhs) [inline amr-wind/core/field_ops.H:390]:
```suggestion
                    if (mag > eps) {
```

File: amr-wind/core/field_ops.H (hunk 2 above)
Comment (gdeskos) [reply amr-wind/core/field_ops.H:342]: I am working on a unit test now. this is the same idea as in incflo_compute_dt.cpp --> https://github.com/Exawind/amr-wind/blob/569d308a9b788634035236c20143c4a31645d660/amr-wind/incflo_compute_dt.cpp#L52-#L64

**Pull request merged**